src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py

```diff
@@ -15066,35 +15066,200 @@ def stack(
             .melt(
-                id_vars=index_names,
+                id_vars=index_cols,
                 value_vars=self.columns,
-                var_name="index_second_level",
+                var_name=col_label,
                 value_name=MODIN_UNNAMED_SERIES_LABEL,
                 ignore_index=False,
             )
             .sort_rows_by_column_values(
-                columns=index_names,  # type: ignore
+                columns=index_cols,  # type: ignore
                 ascending=[True],
                 kind="stable",
                 na_position="last",
                 ignore_index=False,
             )
+            # TODO: SNOW-1524695: Fix `NULL_REPLACE` values in output of `melt`
             .replace(to_replace=UNPIVOT_NULL_REPLACE_VALUE, value=np.nan)
-            .set_index_from_columns(index_names + ["index_second_level"])  # type: ignore
-            .set_index_names([None, None])
+            .set_index_from_columns(index_cols + [col_label])  # type: ignore
         )
 
+        index_cols_replace_none = [
+            None if index_cols[i] == "index" else index_cols[i]
+            for i in range(len(index_cols))
+        ]
+        # Set the correct index names based on column and index names
+        if self.columns.name is None and not does_index_names_contain_none:
+            qc = qc.set_index_names(index_cols_replace_none + [None])  # type: ignore
+        elif self.columns.name is None and does_index_names_contain_none:
+            qc = qc.set_index_names([None, None])
+        elif does_index_names_contain_none:
+            qc = qc.set_index_names([col_label, None])
+
         if dropna:
             return qc.dropna(axis=0, how="any", thresh=None)
         else:
             return qc
 
+    def unstack(
+        self,
+        level: Union[int, str, list] = -1,
+        fill_value: Optional[Union[int, str, dict]] = None,
+        sort: bool = True,
+        is_series_input: bool = False,
+    ) -> "SnowflakeQueryCompiler":
+        """
+        Pivot a level of the (necessarily hierarchical) index labels.
+
+        Returns a DataFrame having a new level of column labels whose
+        inner-most level consists of the pivoted index labels.
+
+        If the index is not a MultiIndex, the output will be a Series
+        (the analogue of stack when the columns are not a MultiIndex).
+
+        Parameters
+        ----------
+        level : int, str, list, default -1
+            Level(s) of index to unstack, can pass level name.
+
+        fillna : int, str, dict, optional
+            Replace NaN with this value if the unstack produces missing values.
+
+        sort : bool, default True
+            Sort the level(s) in the resulting MultiIndex columns.
+
+        is_series_input : bool, default False
+            Whether the input is a Series, in which case we call `droplevel`
+        """
+        if not isinstance(level, int):
+            ErrorMessage.not_implemented(
```

Comment:
do we have a ticket to support level as a string or list? especially with string, seems something could be used with multiIndex commonly, and we do handle len(self.index_names)> 1. 
we have function parse_levels_to_integer_levels and get_pandas_labels_for_levels and get_snowflake_identifiers_for_levels to help handle the level parameters in general

Reply:
There is a ticket I filed for it -- https://snowflakecomputing.atlassian.net/browse/SNOW-1558364. I added a TODO here and will do this as a follow-up to this PR. I'll try to support multiindex column names in the same follow-up PR as well.